Flaky-test: GeoReplicationTest.testTopicReplication

<!--- 

Instructions for reporting a flaky test using this issue template:

1. Replace test_class in title and body with the short test class name WITHOUT the package name.
2. Replace test_method in title and body with the test method that failed. Multiple methods are flaky, remove the content that refers to the test method.
3. Replace "url here" with a url to an example failure. In the Github Actions workflow run logs, you can right click on the line number to copy a link to the line. Example of such url is https://github.com/apache/pulsar/pull/8892/checks?check_run_id=1531075794#step:9:377 . The logs are available for a limited amount of time (usually for a few weeks).
4. Replace "relevant parts of the exception stacktrace here" with the a few lines of the stack trace that shows at least the exception message and the line of test code where the stacktrace occurred.
5. Replace "full exception stacktrace here" with the full exception stacktrace from logs. This section will be hidden by default.
6. Remove all unused fields / content to unclutter the reported issue. Remove this comment too.

-->
GeoReplicationTest.testTopicReplication is flaky. It fails sporadically.

[example failure](https://github.com/apache/pulsar/runs/6715214224?check_suite_focus=true#step:12:15869)

```
  Error:  Tests run: 43, Failures: 6, Errors: 0, Skipped: 20, Time elapsed: 552.812 s <<< FAILURE! - in TestSuite
  Error:  testTopicReplication(org.apache.pulsar.tests.integration.messaging.GeoReplicationTest)  Time elapsed: 18.618 s  <<< FAILURE!
  java.lang.AssertionError: expected object to not be null
  	at org.testng.Assert.fail(Assert.java:99)
  	at org.testng.Assert.assertNotNull(Assert.java:942)
  	at org.testng.Assert.assertNotNull(Assert.java:926)
  	at org.apache.pulsar.tests.integration.messaging.GeoReplicationTest.testTopicReplication(GeoReplicationTest.java:124)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
  	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:833)
```

Another occurrence of this flaky test.

**Example failure:** https://github.com/apache/pulsar/actions/runs/27804576061/job/82585497783

**Related PR:** https://github.com/apache/pulsar/pull/26056

**Failed check:** `CI - Integration - Messaging`

**Failed variant:** `testTopicReplication[4](non-persistent)` (the previous `non-persistent` retry attempt was recorded as skipped after the same assertion failure)

**Stack trace:**
```text
java.lang.AssertionError: expected object to not be null
    at org.testng.Assert.assertNotNull(Assert.java:1496)
    at org.testng.Assert.assertNotNull(Assert.java:1480)
    at org.apache.pulsar.tests.integration.messaging.GeoReplicationTest.testTopicReplication(GeoReplicationTest.java:133)
```

**Analysis:** The persistent variant passed, while the non-persistent variant timed out waiting for a message from the remote cluster. The test sends messages immediately after creating the producer and consumer. For non-persistent topics, replication can drop messages when the remote replicator producer is not started or writable yet, so this looks like a startup race where the test sends before the non-persistent replication path is ready.

--- a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/messaging/GeoReplicationTest.java
+++ b/tests/integration/src/test/java/org/apache/pulsar/tests/integration/messaging/GeoReplicationTest.java
@@ -30,6 +30,8 @@ import org.apache.pulsar.client.api.Consumer;
 import org.apache.pulsar.client.api.Message;
 import org.apache.pulsar.client.api.Producer;
 import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.common.naming.TopicName;
+import org.apache.pulsar.common.policies.data.ReplicatorStats;
 import org.apache.pulsar.tests.integration.topologies.PulsarClusterSpec;
 import org.apache.pulsar.tests.integration.topologies.PulsarGeoClusterTestBase;
 import org.awaitility.Awaitility;
@@ -43,6 +45,9 @@ import org.testng.annotations.Test;
  */
 @CustomLog
 public class GeoReplicationTest extends PulsarGeoClusterTestBase {
+
+    private static final int PARTITION_COUNT = 10;
+    private static final int MESSAGES = 10;
 
     @BeforeClass(alwaysRun = true)
     public final void setupBeforeClass() throws Exception {
@@ -68,7 +73,7 @@ public class GeoReplicationTest extends PulsarGeoClusterTestBase {
         cleanup();
     }
 
-    @Test(timeOut = 1000 * 30, dataProvider = "TopicDomain")
+    @Test(timeOut = 1000 * 60, dataProvider = "TopicDomain")
     public void testTopicReplication(String domain) throws Exception {
         String cluster1 = getGeoCluster().getClusters()[0].getClusterName();
         String cluster2 = getGeoCluster().getClusters()[1].getClusterName();
@@ -82,12 +87,12 @@ public class GeoReplicationTest extends PulsarGeoClusterTestBase {
         String topic = domain + "://public/default/testTopicReplication-" + UUID.randomUUID();
         Awaitility.await().atMost(10, TimeUnit.SECONDS).untilAsserted(() -> {
             try {
-                admin.topics().createPartitionedTopic(topic, 10);
+                admin.topics().createPartitionedTopic(topic, PARTITION_COUNT);
             } catch (Exception e) {
                 log.error().attr("topic", topic).exception(e).log("Failed to create partitioned topic .");
                 Assert.fail("Failed to create partitioned topic " + topic);
             }
-            Assert.assertEquals(admin.topics().getPartitionedTopicMetadata(topic).partitions, 10);
+            Assert.assertEquals(admin.topics().getPartitionedTopicMetadata(topic).partitions, PARTITION_COUNT);
         });
         log.info().attr("topic", topic).log("Test geo-replication produce and consume for topic .");
 
@@ -120,7 +125,11 @@ public class GeoReplicationTest extends PulsarGeoClusterTestBase {
                 .attr("topic", topic)
                 .log("Successfully create consumer in cluster for topic .");
 
-        for (int i = 0; i < 10; i++) {
+        if ("non-persistent".equals(domain)) {
+            waitForNonPersistentReplicators(admin, topic, cluster2);
+        }
+
+        for (int i = 0; i < MESSAGES; i++) {
             p.send(String.format("Message [%d]", i).getBytes(StandardCharsets.UTF_8));
         }
         log.info()
@@ -128,7 +137,7 @@ public class GeoReplicationTest extends PulsarGeoClusterTestBase {
                 .attr("topic", topic)
                 .log("Successfully produce message to cluster for topic .");
 
-        for (int i = 0; i < 10; i++) {
+        for (int i = 0; i < MESSAGES; i++) {
             Message<byte[]> message = c.receive(10, TimeUnit.SECONDS);
             Assert.assertNotNull(message);
         }
@@ -136,5 +145,21 @@ public class GeoReplicationTest extends PulsarGeoClusterTestBase {
                 .attr("cluster", cluster2)
                 .attr("topic", topic)
                 .log("Successfully consume message from cluster for topic .");
+    }
+
+    private void waitForNonPersistentReplicators(PulsarAdmin admin, String topic, String remoteCluster)
+            throws Exception {
+        TopicName topicName = TopicName.get(topic);
+        Awaitility.await().atMost(20, TimeUnit.SECONDS).untilAsserted(() -> {
+            for (int i = 0; i < PARTITION_COUNT; i++) {
+                String partitionName = topicName.getPartition(i).toString();
+                ReplicatorStats replicatorStats = admin.topics()
+                        .getStats(partitionName)
+                        .getReplication()
+                        .get(remoteCluster);
+                Assert.assertNotNull(replicatorStats);
+                Assert.assertTrue(replicatorStats.isConnected());
+            }
+        });
     }
 }

